crates/toml/src/de.rs

```diff
@@ -38,9 +38,9 @@
 /// assert_eq!(config.owner.name, "Lisa");
 /// ```
 #[cfg(feature = "parse")]
-pub fn from_str<T>(s: &'_ str) -> Result<T, Error>
+pub fn from_str<'a, T>(s: &'a str) -> Result<T, Error>
```

Comment:
Please add a test for borrowed data to ensure this works.

If it does work, then more investigation is needed because it shouldn't [according to the serde docs](https://serde.rs/lifetimes.html#trait-bounds).  `&str` gets parsed into owned types and then we deserialize that.

Reply:
I've just added the test in https://github.com/toml-rs/toml/pull/505/commits/44563c2879b8d1b73caa17b9e43d216e3737948e, thanks!

I think that the tested code works because `String`s, being owned types, have a implicit `'static` lifetime bound (they can live indefinitely long until dropped). `Cow`s support storing owned data, so a valid lifetime for a `Cow` that either stores owned data or static strings is `Cow<'static, str>`.  Due to [lifetime coercion](https://doc.rust-lang.org/rust-by-example/scope/lifetime/lifetime_coercion.html), a `'static` lifetime can be used whenever a shorter lifetime is expected, such as `'de` in this case: the lifetime of the serialized data source.

Reply:
Is that sufficient if the user explicitly declared the field as `borrow`?

What about if the user isn't using `Cow`?

Reply:
First, this isn't a morality discussion.

Second, we are shifting a compile time error into a runtime error for the easy path while, as you pointed out, a workaround is available for those that want to borrow.  In API design, we should be emphasizing the safe path with the easy path while ideally still supporting the unsafe route.  The current solutions fits that to a T.  I don't know why the serde docs make the recommendation they do but in this case, I see it being useful.

I'm going to go ahead and close this and ask any further discussion to happen in the appropriate issue (#490).